src/oom/oomd-util.c

```diff
@@ -0,0 +1,352 @@
+        ResourcePressure curr_mem_pressure = {0, 0, 0, 0};
+        char *key, *target = NULL;
+        OomdCGroupContext *ctx;
+        usec_t diff;
+        Iterator i;
+
+        assert(h);
+        assert(last_hit_memory_threshold);
+
+        HASHMAP_FOREACH_KEY(ctx, key, h, i)
+                /* Do a weighted comparison (we care more about 10s, then 60s, then 300s) */
+                if (ctx->memory_pressure.avg10 * 3 + ctx->memory_pressure.avg60 * 2 + ctx->memory_pressure.avg300 >
+                                curr_mem_pressure.avg10 * 3 + curr_mem_pressure.avg60 * 2 + curr_mem_pressure.avg300) {
+                        curr_mem_pressure = ctx->memory_pressure;
+                        target = key;
+                }
+
+        if (curr_mem_pressure.avg10 > threshold) {
+                if (*last_hit_memory_threshold == 0)
+                        *last_hit_memory_threshold = now(CLOCK_REALTIME);
+
+                diff = now(CLOCK_REALTIME) - *last_hit_memory_threshold;
+                if (diff >= duration) {
+                        log_debug("Memory pressure for %s is greater than %.1f for more than %"PRIu64" seconds",
+                                        target, threshold, duration / USEC_PER_SEC);
+                        return target;
+                }
+        } else
+                *last_hit_memory_threshold = 0;
+
+        return NULL;
+}
+
+bool oomd_swap_free_below(const OomdSystemContext ctx, unsigned threshold_percent) {
+        uint64_t swap_threshold;
+        bool ret;
+
+        assert(threshold_percent <= 100);
+
+        swap_threshold = ctx.swap_total * ((uint64_t) threshold_percent) / 100;
+        ret = (ctx.swap_total - ctx.swap_used) < swap_threshold;
+        if (ret)
+                log_debug("Swap total (%"PRIu64") - used (%"PRIu64") is less than %u%%",
+                                ctx.swap_total, ctx.swap_used, threshold_percent);
+        return ret;
+}
+
+int compare_pgscan(const OomdCGroupContext *c1, const OomdCGroupContext *c2) {
+        assert(c1);
+        assert(c2);
+
+        if (c1->pgscan > c2->pgscan)
+                return 1;
+        else if (c1->pgscan < c2->pgscan)
+                return -1;
+        else
+                return 0;
+}
+
+int compare_swap_usage(const OomdCGroupContext *c1, const OomdCGroupContext *c2) {
+        assert(c1);
+        assert(c2);
+
+        if (c1->swap_usage > c2->swap_usage)
+                return 1;
+        else if (c1->swap_usage < c2->swap_usage)
+                return -1;
+        else
+                return 0;
+}
+
+int oomd_sort_cgroups(Hashmap *h, compare_func_t compare_func, const char *prefix, char ***ret) {
+        _cleanup_free_ char **sorted_cgroups = NULL;
+        OomdCGroupContext *c1, *c2;
+        Iterator i;
+        char *key;
+
+        assert(h);
+        assert(compare_func);
+        assert(ret);
+
+        HASHMAP_FOREACH_KEY(c1, key, h, i)
+                if (prefix == NULL || startswith(key, prefix)) {
```

Comment:
in intereste of keeping indentation levels small, maybe invert condition and do continue early (like in that other case i commented on)